astropy/time/formats.py

```diff
@@ -141,11 +142,42 @@ def scale(self):
     def scale(self, val):
         self._scale = val
 
+    def mask_if_needed(self, value):
+        if self.masked:
+            value = np.ma.array(value, mask=self.mask, copy=False)
+        return value
+
+    @property
+    def mask(self):
+        if 'mask' not in self.cache:
+            self.cache['mask'] = np.isnan(self.jd2)
+            if self.cache['mask'].shape:
+                self.cache['mask'].flags.writeable = False
+        return self.cache['mask']
+
+    @property
+    def masked(self):
+        if 'masked' not in self.cache:
+            self.cache['masked'] = bool(np.any(self.mask))
+        return self.cache['masked']
+
+    @property
+    def jd2_filled(self):
+        return np.nan_to_num(self.jd2) if self.masked else self.jd2
+
+    @lazyproperty
+    def cache(self):
+        """
+        Return the cache associated with this instance.
+        """
+        return defaultdict(dict)
+
     def _check_val_type(self, val1, val2):
         """Input value validation, typically overridden by derived classes"""
-        if not (val1.dtype == np.double and np.all(np.isfinite(val1)) and
-                (val2 is None or
-                 val2.dtype == np.double and np.all(np.isfinite(val2)))):
+        # val1 cannot contain nan, but val2 can contain nan
+        ok1 = val1.dtype == np.double and np.all(np.isfinite(val1))
```

Comment:
Why ``isfinite`` and not ``isnan``? This will catch inf/-inf

Reply:
This line for `ok1` is the same as it has always been and says that the user input must be double and finite and not NaN.  (`isfinite(nan) is False`).

The next line for `ok2` is mostly the same but requires that all values be a finite number or NaN.  (`isinf(nan) is False`).

Summary: same as before except that NaN is allowed for `val2`.

Reply:
Ok sounds good!